### ISSUE TITLE
[BUG?]The same code behaves differently on pytorch based on py2 and py3 whose version are both 0.3.0.post_4

### ISSUE THREAD

**Issue by sonack** [pytorch/pytorch#5870]
I have tried to train an autoencoder-like network with pytorch based on py3, its loss decreased normally on my local machine. However, when I moved the code to my training server whose pytorch is based on py2 , I found the loss was never converge. I debugged it for a long time, and suspected it's because of the framework itself finally, because when I changed the server's environment to pytorch based on py3, the bug disappeared and behaved the same as my local machine, the loss could be declined quickly.

All the training data and code are the same, following is my network's definition and training code.

In summary, I found the bug appear's version is **0.3.0.post4 of python2**.
I found the bug disappear's version is **0.3.0.post4 of python3(my local machine)**, and **'0.3.1.post2 of python3(my server's new environment)**.

**trainning code**
```python
#coding:utf8
from __future__ import print_function
from config import opt
import os
import numpy as np
import math
import torch as t
import models

from torch.utils.data import DataLoader
from torch.autograd import Variable
from torchnet import meter
from utils import AverageValueMeter
from torchvision import transforms, datasets
import pdb

from utils.visualize import Visualizer, PlotSaver
from extend import RateLoss, LimuRateLoss
import torch.backends.cudnn as cudnn
import time


def train(**kwargs):
    opt.parse(kwargs)
    ps = PlotSaver("TrainImageNet12WithoutImpMap_"+time.strftime("%m_%d_%H:%M:%S")+".log")


    # step1: Model
    model = getattr(models, opt.model)(use_imp = opt.use_imp, model_name="CWCNN_limu_ImageNet_imp_r={r}_γ={w}".format(
                                                                r=opt.rate_loss_threshold,
                                                                w=opt.rate_loss_weight)
                                                                if opt.use_imp else None)

    if opt.use_gpu:
        model.cuda()

   
    normalize = transforms.Normalize(
                mean=[0.5, 0.5, 0.5],
                std=[0.5, 0.5, 0.5]
                )

    train_data_transforms = transforms.Compose(
        [
            transforms.ToTensor(),
            normalize
        ]
    )
    val_data_transforms = transforms.Compose(
        [
            transforms.ToTensor(),
            normalize
        ]
    )

    train_data = datasets.ImageFolder(
        opt.train_data_root,
        train_data_transforms
    )
    val_data = datasets.ImageFolder(
        opt.val_data_root,
        val_data_transforms
    )
 
    train_dataloader = DataLoader(train_data, opt.batch_size, shuffle=True, num_workers=opt.num_workers)
    val_dataloader = DataLoader(val_data, opt.batch_size, shuffle=False, num_workers=opt.num_workers)

    # step3: criterion and optimizer

    mse_loss = t.nn.MSELoss(size_average = False)

    lr = opt.lr

    optimizer = t.optim.Adam(model.parameters(), lr=lr, betas=(0.9, 0.999))

    start_epoch = 0

    # step4: meters
    mse_loss_meter = AverageValueMeter()
  
    previous_loss = 1e100
    tolerant_now = 0
    same_lr_epoch = 0


    for epoch in range(start_epoch+1, opt.max_epoch+1):
        same_lr_epoch += 1
  
        mse_loss_meter.reset()
        total_loss_meter = mse_loss_meter
        model.train()

        for idx, (data, _) in enumerate(train_dataloader):
            ipt = Variable(data)

            if opt.use_gpu:
                ipt = ipt.cuda()

            optimizer.zero_grad()  
            reconstructed = model(ipt)

            loss = mse_loss(reconstructed, ipt)
            caffe_loss = loss / (2 * opt.batch_size)

            total_loss = caffe_loss

            total_loss.backward()
            optimizer.step()

            mse_loss_meter.add(caffe_loss.data[0])

            if idx % opt.print_freq == opt.print_freq - 1:
                pass    # print sth. I omitted it for clarity

       
        model.save(optimizer, epoch)

        # valdation 
        val()

        if opt.use_early_adjust:
            if total_loss_meter.value()[0] > previous_loss:
                tolerant_now += 1
                if tolerant_now == opt.tolerant_max:
                    tolerant_now = 0
                    same_lr_epoch = 0
                    lr = lr * opt.lr_decay
                    for param_group in optimizer.param_groups:
                        param_group['lr'] = lr
                    print ('Due to early stop anneal lr to',lr,'at epoch',epoch)
            else:
                tolerant_now -= 1

        if same_lr_epoch and same_lr_epoch % opt.lr_anneal_epochs == 0:
            same_lr_epoch = 0
            tolerant_now = 0
            lr = lr * opt.lr_decay
            for param_group in optimizer.param_groups:
                param_group['lr'] = lr
            print ('Due to full epochs anneal lr to',lr,'at epoch',epoch)

        previous_loss = total_loss_meter.value()[0]


```
 
**Model definition**

```python
#coding:utf-8
from .BasicModule import BasicModule
from torch import nn
from torch.nn import functional as F
import sys
import math


from extend import Round
from torch.nn.init import xavier_uniform

import pdb

class ResidualBlock(nn.Module):
    '''
    Residual Module
        input(nchw) -> Conv(3x3x128, pad 1) -> ReLu -> Conv(3x3xc, pad 1) -> ReLU
        input(nchw)                                                                +  = output(nchw)
    '''
    def __init__(self, ch_in, ch_out, shortcut = None):
        super(ResidualBlock, self).__init__()
        self.left = nn.Sequential(
            nn.Conv2d(ch_in, 128, 3, 1, 1),  # ch_in, ch_out, kernel_size, stride, pad
            nn.ReLU(inplace=False),
            nn.Conv2d(128, ch_out, 3, 1, 1),
            nn.ReLU(inplace=False),
        )
        self.right = shortcut
    
    def forward(self, x):
        out = self.left(x)
        residual = x if self.right is None else self.right(x)
        out += residual
        return F.relu(out, inplace=False)


def xavier_init(data):
    fan_in = data.numel() / data.size(0)
    scale = math.sqrt(3 / fan_in)
    data.uniform_(-scale, scale)


def weights_initialization(m):
    classname = m.__class__.__name__
    if classname.find('Conv') != -1:
        xavier_init(m.weight.data)
        m.bias.data.fill_(0)


# 
class ContentWeightedCNN(BasicModule):
    '''
    BasicModule is just a Module with save() and load() convenient method, I don't think it matters, so I ommited the relevant files.
    '''
    def __init__(self, use_imp = True, model_name = None):
        super(ContentWeightedCNN, self).__init__()
        self.model_name = model_name if model_name else 'ContentWeightedCNN'
        self.encoder = self.make_encoder()
        self.decoder = self.make_decoder()
        self.reset_parameters()
    
    def reset_parameters(self):
        self.apply(weights_initialization)
    


    def forward(self, x):
        mgdata = self.encoder(x)
        enc_data = mgdata
        dec_data = self.decoder(enc_data)
        return dec_data

    def make_encoder(self):
        layers = [
            nn.Conv2d(3, 128, 8, 4, 2),
            nn.ReLU(inplace=False), # 54

            ResidualBlock(128, 128),

            nn.Conv2d(128, 256, 4, 2, 1), # 115
            nn.ReLU(inplace=False),

            ResidualBlock(256, 256),

            nn.Conv2d(256, 256, 3, 1, 1), #192
            nn.ReLU(inplace=False),

            ResidualBlock(256, 256),

            nn.Conv2d(256, 64, 1, 1, 0),   
            nn.Sigmoid(),                    
        ]
        return nn.Sequential(*layers)



    def make_decoder(self):
        layers = [
            nn.Conv2d(64, 512, 3, 1, 1),
            nn.ReLU(inplace=False),

            ResidualBlock(512, 512),

            nn.Conv2d(512, 512, 3, 1, 1),
            nn.ReLU(inplace=False),

            ResidualBlock(512, 512),

            nn.PixelShuffle(2),

            nn.Conv2d(128, 256, 3, 1, 1),
            nn.ReLU(inplace=False),

            ResidualBlock(256, 256),

            nn.PixelShuffle(4),

            nn.Conv2d(16, 32, 3, 1, 1),
            nn.ReLU(inplace=False),
            
            nn.Conv2d(32, 3, 1, 1, 0)   
        ]
        return nn.Sequential(*layers)    
```

For clarity, I omitted some code.
If you need more relevant code to understand or reproduce, comment to tell me.

I know it is a little bit long, but I don't know where is the matter. Thanks for your attention.

**Comment by apaszke** [comment]
Please try to reduce the snippet. It’s a lot of work for us to try to understand all the code that’s posted in the issues. Keep deleting code as long as your model trains in Python 3 and fails to do so in Python 2. Once you can’t delete anything anymore please let us know.

**Comment by sonack** [comment]
@apaszke  OK, but I don't have enough time to do it now. I will close this issue until someone else find something similar.

**Comment by SsnL** [comment]
from a rough scan, this looks like that it might be the culprit:
```
    scale = math.sqrt(3 / fan_in)
```
in py2, `int / int` is an `int`; in py3, it is a `float`.

**Comment by sonack** [comment]
@SsnL  Thank you very much! I will try it.
